src/util/server_table.c

```diff
@@ -42,53 +47,62 @@ int server_table_add(charbuf server_id, uint64_t handle)
     return MEM_ALLOC_FAIL;
   }
 
-  tmp_entry.id = new_charbuf(server_id.len);
-  if (server_id.len != tmp_entry.id.len)
-  {
-    pelz_log(LOG_ERR, "Charbuf creation error.");
-    return ERR_BUF;
-  }
-
-  memcpy(tmp_entry.id.chars, server_id.chars, tmp_entry.id.len);
   data_size = retrieve_from_unseal_table(handle, &data);
   if (data_size == 0)
   {
     pelz_log(LOG_ERR, "Failure to retrive data from unseal table.");
     return RET_FAIL;
   }
 
-  tmp_entry.value.cert = new_charbuf(data_size);
-  if (data_size != tmp_entry.value.cert.len)
+  ret = unmarshal_ec_der_to_x509(&data, &data_size, &tmp_entry.value.cert);
+  if (ret)
   {
-    pelz_log(LOG_ERR, "Charbuf creation error.");
-    return ERR_BUF;
+    pelz_log(LOG_ERR, "Unmarshal DER to X509 Failure");
+    free(data);
+    return ERR_X509;
   }
+  free(data);
 
-  memcpy(tmp_entry.value.cert.chars, data, tmp_entry.value.cert.len);
+  X509_NAME *subj = X509_get_subject_name(tmp_entry.value.cert);
 
-  if (table_lookup(SERVER, tmp_entry.id, &index) == 0)
+  for (;;)
   {
-    tmpcert = new_charbuf(server_table.entries[index].value.cert.len);
-    if (server_table.entries[index].value.cert.len != tmpcert.len)
+    int count = X509_NAME_get_index_by_NID(subj, NID_commonName, lastpos);
+
+    if (count == -1)
     {
-      pelz_log(LOG_ERR, "Charbuf creation error.");
-      return (1);
+      break;
     }
-    memcpy(tmpcert.chars, server_table.entries[index].value.cert.chars, tmpcert.len);
-    if (cmp_charbuf(tmpcert, tmp_entry.value.cert) == 0)
+    lastpos = count;
+  }
+  X509_NAME_ENTRY *entry = X509_NAME_get_entry(subj, lastpos);
```

Comment:
I think you need some additional error checking of ```X509_NAME_get_index_by_NID``` returns -1 the first time it's called. If my understanding is right that would mean ```subj``` doesn't contain a valid ```Name``` field, so we shouldn't then try to extract the name.

Reply:
With the check that ```subj``` is a valid pointer, ```X509_NAME_get_entry``` must have at least one valid field.  The for loop ends when ```X509_NAME_get_index_by_NID``` returns -1 but ```lastpos``` has the last valid value from the count which will be the valid 0 position if  ```X509_NAME_get_index_by_NID``` returns -1 on the first loop.